### ISSUE TITLE
Site Lisp, wrappers, and achieving consistency with `invocation-directory`, `invocation-name`, and `process-environment`

### ISSUE THREAD

**Issue by psionic-k** [nix-community/emacs-overlay#388]
Elpaca likes to calculate the exact Emacs binary by using `(concat invocation-directory invocation-name)`  This is an accurate way to avoid getting the wrong Emacs version for compiling packages in a sub-process.

However, it preserves the `process-environment` which contains `EMACSLOADPATH` and other variables.  The result is that this direct invocation of the Emacs binary goes looking for an unexpected site lisp location, in the middle wrapper's derivation.

By "middle wrapper" I'm referring to an Emacs obtained like so:

```nix
    programs.emacs = {
      package = (pkgs.emacsPackagesFor pkgs.emacs-unstable).emacsWithPackages
        (epkgs: [epkgs.treesit-grammars.with-all-grammars]);
    };
```

When calling the final binary without the wrappers but with the process environment, it complains about not finding site lisp.  Here's an error message:

```
  [2024-02-03 14:33:46] Byte compiling
  [2024-02-03 14:33:46] $/nix/store/83csl1ckkydzwqdaj20dmpyfl1sfgyym-emacs-unstable-29.2/bin/emacs -q --batch --eval (let ((gc-cons-percentage 1.0) (native-comp-eln-load-path '("/home/satoshi/Desktop/test/eln-cache/" "/home/satoshi/.nix-profile/share/emacs/native-lisp/" "/nix/profile/share/emacs/native-lisp/" "/home/satoshi/.local/state/nix/profile/share/emacs/native-lisp/" "/etc/profiles/per-user/satoshi/share/emacs/native-lisp/" "/nix/var/nix/profiles/default/share/emacs/native-lisp/" "/run/current-system/sw/share/emacs/native-lisp/" "/nix/store/p8ik8k2s96xh2y8ibaq4zwr36m9n3q1s-emacs-packages-deps/share/emacs/native-lisp" "/nix/store/p5pbh8jc1k1kfqhi25ldg4mdfclpdn23-emacs-packages-deps/share/emacs/native-lisp" "/nix/store/83csl1ckkydzwqdaj20dmpyfl1sfgyym-emacs-unstable-29.2/lib/emacs/29.2/native-lisp/"))) (dolist (dir '("/home/satoshi/Desktop/test/elpaca/builds/orderless")) (let ((default-directory dir)) (add-to-list 'load-path dir) (normal-top-level-add-subdirs-to-load-path))) (byte-recompile-directory "/home/satoshi/Desktop/test/elpaca/builds/orderless" 0 'force))
  [2024-02-03 14:33:46] Cannot open load file: No such file or directory, /nix/store/j0mwkna8j7163clyjnydilx3skl0jyr3-emacs-unstable-with-packages-29.2/share/emacs/site-lisp/site-start
```

Note, the Emacs in `/nix/store/83csl1ckkydzwqdaj20dmpyfl1sfgyym-emacs-unstable-29.2` is looking for the site lisp in `/nix/store/j0mwkna8j7163clyjnydilx3skl0jyr3-emacs-unstable-with-packages-29.2/`

It feels like either the wrapper should re-use the process environment correctly or it should modify `invocation-directory` and `invocation-name` because that's not a bad way to find the correct executable when doing what Elpaca is doing, calling Emacs as a subprocess.

**Comment by psionic-k** [comment]
Looks like DOOM's installer tries to do the same `(concat invocation-directory invocation-name)` thing.

It would probably be sufficient to add expressions to propagate a correct `invocation-directory` and `invocation-name` through the wrappers so that whatever Emacs is actually called can properly self-reference from within Elisp programs.

1. Introduce `INVOCATION_DIRECTORY` variable into the wrappers.. I can't tell how these are created
2. Patch an Elisp file to overwrite `invocation-directory` if the value is not empty

**Comment by jian-lin** [comment]
This should be an issue in Nixpkgs instead of this overlay.

Actually, in [site-start][1], the entry added by our wrapper is removed.  The [motivation][2] of that logic seems to solve the very problem you have.

And I cannot reproduce you problem:
- run `M-& /nix/store/mzdvpxzkhlryxmbxwbih7bgvkzkw1ly7-emacs-29.1/bin/emacs` in a wrapped Emacs
- there is no entries added by our wrapper in the `load-path` of the subprocess unwrapped Emacs

[1]: https://github.com/NixOS/nixpkgs/blob/d6ed78b53c149c5b3dfea8aefb6ecb8795a05790/pkgs/applications/editors/emacs/site-start.el#L20-L30
[2]: https://github.com/NixOS/nixpkgs/commit/23d4bfb6661ca57a9e331a2cf4184232d38ac38b

**Comment by psionic-k** [comment]
Thanks for checking.

To reproduce, in a wrapped Emacs, evaluate:
```elisp
;; (concat invocation-directory invocation-name)
;; "/nix/store/83csl1ckkydzwqdaj20dmpyfl1sfgyym-emacs-unstable-29.2/bin/emacs"

(start-process "child-emacs" "emacs" (concat invocation-directory invocation-name))
```
Note that the error indicates a different store path:
```
Loading /home/satoshi/.emacs.d/posimacs/posimacs-early-init.el (source)...done
byte-code: Cannot open load file: No such file or directory, /nix/store/j0mwkna8j7163clyjnydilx3skl0jyr3-emacs-unstable-with-packages-29.2/share/emacs/site-lisp/site-start
```

> I cannot reproduce you problem
> /nix/store/mzdvpxzkhlryxmbxwbih7bgvkzkw1ly7-emacs-29.1/bin/emacs

The wrapper will run fine when invoked directly.  However, the use of `(concat invocation-path invocation-name)` will try to load the Emacs after the wrapper.  This one fails when run without the wrapper.

My Emacs is constructed as:
```
    programs.emacs = {
      package = (pkgs.emacsPackagesFor pkgs.emacs-unstable).emacsWithPackages
        (epkgs: [epkgs.treesit-grammars.with-all-grammars]);
    };
```
I don't know why, but this creates three derivations, where two are wrappers.  The error path involves the middle wrapper's derivation path.

My versions are all pretty new, 23.11 and `emacs-overlay` from within the last month.

**Comment by jian-lin** [comment]
> Thanks for checking.
> 
> To reproduce, in a wrapped Emacs, evaluate:
> 
> ```emacs-lisp
> ;; (concat invocation-directory invocation-name)
> ;; "/nix/store/83csl1ckkydzwqdaj20dmpyfl1sfgyym-emacs-unstable-29.2/bin/emacs"
> 
> (start-process "child-emacs" "emacs" (concat invocation-directory invocation-name))
> ```

I still cannot reproduce it.

> 
> Note that the error indicates a different store path:
> 
> ```
> Loading /home/satoshi/.emacs.d/posimacs/posimacs-early-init.el (source)...done
> byte-code: Cannot open load file: No such file or directory, /nix/store/j0mwkna8j7163clyjnydilx3skl0jyr3-emacs-unstable-with-packages-29.2/share/emacs/site-lisp/site-start
> ```
> 
> > I cannot reproduce you problem
> > /nix/store/mzdvpxzkhlryxmbxwbih7bgvkzkw1ly7-emacs-29.1/bin/emacs
> 
> The wrapper will run fine when invoked directly. However, the use of `(concat invocation-path invocation-name)` will try to load the Emacs after the wrapper. This one fails when run without the wrapper.
> 
> My Emacs is constructed as:
> 
> ```
>     programs.emacs = {
>       package = (pkgs.emacsPackagesFor pkgs.emacs-unstable).emacsWithPackages
>         (epkgs: [epkgs.treesit-grammars.with-all-grammars]);
>     };
> ```
> 
> I don't know why, but this creates three derivations, where two are wrappers. The error path involves the middle wrapper's derivation path.

Now I know why I cannot reproduce it.  Your Emacs is wrapped **twice** by a manual `emacsWithPackages` and home-manager's `programs.emacs`.  Check my [explanation][1] about why this is not a good practice. 

> 
> My versions are all pretty new, 23.11 and `emacs-overlay` from within the last month.

[1]: https://github.com/NixOS/nixpkgs/issues/267548#issuecomment-1833505396

**Comment by psionic-k** [comment]
Removed double wrap and confirmed no remaining issues.  I have deleted my contribution to the cargo culting of that expression.

```nix
    emacs = pkgs.emacs-unstable.pkgs.withPackages
      (epkgs: [epkgs.treesit-grammars.with-all-grammars]);
```

Maybe the README should be overhauled for best practices.  

Just at first glance:
1. "Quick start" first recommends OS level change, which is generally slow moving and too heavy for active use.  Also relies on NixOS rather than just having nix.
2. Use of `services` is neat but since Emacs can run the server on its own during init, I think this should be moved down
3. No mention of `withPackages`. The `pkgs.emacs.pkgs.withPackages` expression I am using now is a little bit unexpected.  Is it even right?
4. Mention home manager for installing persistent Emacs or flakes for many ephemeral Emacs
5. `nix-build` impure usage.  It's possible to make the flake for `emacs-overlay` deliver both the overlay and applications from one repository

Should I draft these changes?

I think we can close this issue.  It's done.
